File: 2025-10-October-GFG-POTD/October 11 - Maximum path sum/Solution.java
/*
class Node{
    int data;
    Node left, right;
    Node(int d){
        data=d;
        left=right=null;
    }
}
*/

class Solution {
    int res; // global variable to store maximum path sum

    int findMaxSum(Node root) {
        res = Integer.MIN_VALUE;
        maxPathDown(root);
        return res;
    }

    int maxPathDown(Node node) {
        if (node == null) return 0;

        // Compute maximum path sum on left and right subtrees
        int left = Math.max(0, maxPathDown(node.left)); 
        int right = Math.max(0, maxPathDown(node.right));

        // Update global result (max path sum through the current node)
        res = Math.max(res, node.data + left + right);

        // Return max path sum going upward (only one branch)
        return node.data + Math.max(left, right);
    }
}

File: 2025-10-October-GFG-POTD/October 11 - Maximum path sum/Solution.py
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None


class Solution:
    def __init__(self):
        self.res = float('-inf')

    def findMaxSum(self, root):
        self.maxPathDown(root)
        return self.res

    def maxPathDown(self, node):
        if node is None:
            return 0

        # Compute maximum path sum on left and right subtrees
        left = max(0, self.maxPathDown(node.left))
        right = max(0, self.maxPathDown(node.right))

        # Update global result (max path sum through the current node)
        self.res = max(self.res, node.data + left + right)

        # Return max path sum going upward (only one branch)
        return node.data + max(left, right)
